source/speech/sayAll.py

```diff
@@ -15,92 +18,278 @@
+
+	def _flushPendingSpeech(self) -> SpeechSequence:
+		"""
+		@return: may be empty sequence
+		"""
+		# Place the last incomplete phrase in to finalSpeechSequence to be spoken now
+		pending = self._pendingSpeechSequence
+		self._pendingSpeechSequence = []
+		return pending
+
+	def _getSpeech(
+			self,
+			speechSequence: SpeechSequence
+	) -> SpeechSequence:
+		"""
+		@return: May be an empty sequence
+		"""
+		finalSpeechSequence: SpeechSequence = []  # To be spoken now
+		pendingSpeechSequence: speechSequence = []  # To be saved off for speaking later
+		# Scan the given speech and place all completed phrases in finalSpeechSequence to be spoken,
+		# And place the final incomplete phrase in pendingSpeechSequence
+		for index in range(len(speechSequence) - 1, -1, -1):
+			item = speechSequence[index]
+			if isinstance(item, str):
+				m = self.re_last_pause.match(item)
+				if m:
+					before, after = m.groups()
+					if after:
+						pendingSpeechSequence.append(after)
+					if before:
+						finalSpeechSequence.extend(self._flushPendingSpeech())
+						finalSpeechSequence.extend(speechSequence[0:index])
+						finalSpeechSequence.append(before)
+						# Apply the last language change to the pending sequence.
+						# This will need to be done for any other speech change commands introduced in future.
+						for changeIndex in range(index - 1, -1, -1):
+							change = speechSequence[changeIndex]
+							if not isinstance(change, LangChangeCommand):
+								continue
+							pendingSpeechSequence.append(change)
+							break
+						break
+				else:
+					pendingSpeechSequence.append(item)
+			else:
+				pendingSpeechSequence.append(item)
+		if pendingSpeechSequence:
+			pendingSpeechSequence.reverse()
+			self._pendingSpeechSequence.extend(pendingSpeechSequence)
+		return finalSpeechSequence
 
-lastSayAllMode = None
-#: The active say all manager.
-#: This is a weakref because the manager should be allowed to die once say all is complete.
-_activeSayAll = lambda: None # Return None when called like a dead weakref.
-
-
-def getSpeechWithoutPauses() -> "speech.SpeechWithoutPauses":
-	"""Returns an instance of `speech.SpeechWithoutPauses` which should be used for say all
-	creating it if necessary."""
-	if getSpeechWithoutPauses.speechWithoutPausesInstance is None:
-		getSpeechWithoutPauses.speechWithoutPausesInstance = speech.SpeechWithoutPauses(speakFunc=speech.speak)
-	return getSpeechWithoutPauses.speechWithoutPausesInstance
-
-
-getSpeechWithoutPauses.speechWithoutPausesInstance: Optional["speech.SpeechWithoutPauses"] = None
-
-
-def stop():
-	active = _activeSayAll()
-	if active:
-		active.stop()
 
-def isRunning():
-	"""Determine whether say all is currently running.
-	@return: C{True} if say all is currently running, C{False} if not.
-	@rtype: bool
-	"""
-	return bool(_activeSayAll())
+class _SayAllHandler:
+	def __init__(self):
+		self.lastSayAllMode = None
+		self.speechWithoutPausesInstance = SpeechWithoutPauses(speakFunc=speak)

```

Comment:
Rather than import speech and create a `SpeechWithoutPauses` instance, could you have it passed into the constructor.

Reply:
speech is still imported for aspects of `SpeechWithoutPauses`, but I have included this change